File: Python/ByTier/E/1767_E_Algebra_Flash.py
# Problem: CF 1767 E - Algebra Flash
# https://codeforces.com/contest/1767/problem/E

"""
Algebra Flash 2.2 - Colored Platforms

This solution uses a Union-Find (Disjoint Set Union) data structure to group platforms by their colors,
then applies dynamic programming with bitmasks to find the minimum cost to activate platforms such
that a path from platform 1 to platform n exists.

Time Complexity: O(n + m^2 + 2^m * m + 2^l * l) where m is the number of colors and l is the size of the largest group
Space Complexity: O(m^2 + 2^m + 2^l)

Algorithms/Techniques:
- Union-Find for grouping connected components
- Bitmask DP for selecting minimal activating cost in groups
- Graph connectivity checking via adjacency matrix represented with bitmask
"""

import io
import os

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v):
    return u * m1 + v


def get_root(s):
    v = []
    while not s == root[s]:
        v.append(s)
        s = root[s]
    for i in v:
        root[i] = s
    return s


def unite(s, t):
    rs, rt = get_root(s), get_root(t)
    if not rs ^ rt:
        return
    if rank[s] == rank[t]:
        rank[rs] += 1
    if rank[s] >= rank[t]:
        root[rt] = rs
        size[rs] += size[rt]
    else:
        root[rs] = rt
        size[rt] += size[rs]
    return


def same(s, t):
    return True if get_root(s) == get_root(t) else False


def get_size(s):
    return size[get_root(s)]


n, m = map(int, input().split())
c = list(map(int, input().split()))
x = [0] + list(map(int, input().split()))
m1 = m + 1
m2 = (m + 1) * (m + 1)
y = [0] * m2
root = [i for i in range(m1)]
rank = [1 for _ in range(m1)]
size = [1 for _ in range(m1)]
for i in range(n - 1):
    u, v = c[i], c[i + 1]
    y[f(u, v)], y[f(v, u)] = 1, 1
    unite(u, v)
y[f(c[0], c[0])], y[f(c[-1], c[-1])] = 1, 1
r = [[] for _ in range(m1)]
for i in range(1, m1):
    r[get_root(i)].append(i)
pow2 = [1]
for _ in range(22):
    pow2.append(2 * pow2[-1])
ans = 0
inf = pow(10, 9) + 1
for r0 in r:
    if not r0:
        continue
    elif len(r0) == 1 and y[f(r0[0], r0[0])]:
        ans += x[r0[0]]
        continue
    a, b = [], []
    for i in range(len(r0)):
        if not i % 2:
            a.append(r0[i])
        else:
            b.append(r0[i])
    m = len(a)
    pm = pow2[m]
    ng = [0] * pm
    for i in range(m):
        ai, pi = a[i], pow2[i]
        for j in range(i, m):
            if y[f(ai, a[j])]:
                ng[pi | pow2[j]] = 1
    for i in range(m):
        pi = pow2[i]
        for j in range(pm):
            if not j & pi:
                ng[j ^ pi] |= ng[j]
    mi = [inf] * pm
    for i in range(pm):
        co = 0
        if ng[i]:
            continue
        for j in range(m):
            if not i & pow2[j]:
                co += x[a[j]]
        mi[i] = co
    for i in range(m):
        pi = pow2[i]
        for j in range(pm - 1, -1, -1):
            if not j & pi:
                mi[j ^ pi] = min(mi[j], mi[j ^ pi])
    l = len(b)
    pl = pow2[l]
    z = [0] * l
    for i in range(l):
        bi, zi = b[i], 0
        for j in range(m):
            if y[f(bi, a[j])]:
                zi ^= pow2[j]
        z[i] = zi
    ng = [0] * pl
    for i in range(l):
        bi, pi = b[i], pow2[i]
        for j in range(i, l):
            if y[f(bi, b[j])]:
                ng[pi | pow2[j]] = 1
    for i in range(l):
        pi = pow2[i]
        for j in range(pl):
            if not j & pi:
                ng[j ^ pi] |= ng[j]
    ans0 = inf
    pm0 = pm - 1
    pl0 = pl - 1
    for i in range(pl):
        co = 0
        if ng[i ^ pl0]:
            continue
        k = 0
        for j in range(l):
            if i & pow2[j]:
                co += x[b[j]]
            else:
                k |= z[j]
        ans0 = min(ans0, co + mi[k ^ pm0])
    ans += ans0
print(ans)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1767/1767_E_Algebra_Flash.cpp
// Problem: CF 1767 E - Algebra Flash
// https://codeforces.com/contest/1767/problem/E

/*
Code Purpose:
This solution solves the "Colored Platforms" problem using a bitmask dynamic programming approach combined with a greedy strategy.
The problem requires finding the minimum cost to activate platforms such that one can traverse from platform 1 to platform n using jumps of size 1 or 2, where all platforms in the path must be activated.

The key idea is to:
1. Identify which colors are required at the start and end.
2. For adjacent platforms, we track which colors can reach each other (i.e., if platform i and i+1 have different colors, then they create a connectivity constraint).
3. Use a recursive function with memoization to determine the maximum number of colors that can be avoided (i.e., not activated) by choosing optimal subsets of colors to activate.
4. Subtract the maximum number of colors not activated from the total cost to get the minimal cost.

Algorithms/Techniques:
- Bitmask Dynamic Programming
- Memoization (via f array)
- Greedy optimization via recursive approach

Time Complexity: O(3^m * m + n), where m is the number of colors and n is the number of platforms.
Space Complexity: O(2^m + m), for storing the memoization table and auxiliary arrays.
*/

#include <stdio.h>
#include <string.h>
#include <algorithm>

using namespace std;
int n, m;
int c[300020];
int x[50];
int v[50];
long long e[50];
int f[1048577]; // Memoization table for bitmask DP

// Recursive function with memoization to compute the maximum number of colors that can be avoided
int F(long long s) {
  // If already computed, return memoized value
  if (s < 1 << 20 && f[s] != -1) {
    return f[s];
  }
  
  // Find the highest color index that is still active in the set
  int p = 63 - __builtin_clzll(s);
  
  // Recursively process without this color
  int r = F(s ^ (1LL << p));
  
  // If this color is not forced to be activated (v[p] == 0),
  // consider activating it and removing all its incompatible neighbors
  if (v[p] == 0) {
    r = max(r, F((s ^ (1LL << p)) & ~e[p]) + x[p]);
  }
  
  // Memoize the result
  if (s < 1 << 20) {
    f[s] = r;
  }
  return r;
}

int main() {
  scanf("%d%d", &n, &m);
  
  // Read platform colors (convert to 0-based indexing)
  for (int i = 1; i <= n; i++) {
    scanf("%d", &c[i]);
    c[i]--;
  }
  
  // Read activation costs
  int z = 0;
  for (int i = 0; i < m; i++) {
    scanf("%d", &x[i]);
    z += x[i]; // Total possible cost
  }
  
  // Mark the colors of start and end platforms as required
  v[c[1]] = 1;
  v[c[n]] = 1;
  
  // Build edge connections between colors of adjacent platforms
  for (int i = 1; i < n; i++) {
    if (c[i] == c[i + 1]) {
      // Same color, mark as required (forced)
      v[c[i]] = 1;
    } else {
      // Different colors, add constraint that they cannot both be inactive
      e[c[i]] |= 1LL << c[i + 1];
      e[c[i + 1]] |= 1LL << c[i];
    }
  }
  
  // Initialize memoization table
  memset(f, -1, sizeof f);
  f[0] = 0; // Base case: no colors selected, max avoided = 0
  
  // Output: Total cost minus maximum number of colors avoided
  printf("%d\n", z - F((1LL << m) - 1));
  
  return 0;
}


// https://github.com/VaHiX/CodeForces/